keras_nlp/models/deberta/disentangled_self_attention.py

```diff
@@ -0,0 +1,376 @@
+        **kwargs,
+    ):
+        super().__init__(name=name, **kwargs)
+
+        # Passed args.
+        self.num_heads = num_heads
+        self.hidden_dim = hidden_dim
+        self.max_position_embeddings = max_position_embeddings
+        self.bucket_size = bucket_size
+        self.dropout = dropout
+
+        # Initializers.
+        self._kernel_initializer = keras.initializers.get(kernel_initializer)
+        self._bias_initializer = keras.initializers.get(bias_initializer)
+
+        # Derived args.
+        self.attn_head_size = hidden_dim // num_heads
+
+        num_type_attn = 3
+        self.scale_factor = 1.0 / math.sqrt(
+            float(num_type_attn * self.attn_head_size)
+        )
+
+        # Layers.
+
+        # Q, K, V linear layers.
+        self._query_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_heads, self.attn_head_size),
+            bias_axes="de",
+            **self._get_common_kwargs_for_sublayer(use_bias=True),
+            name="query",
+        )
+        self._key_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_heads, self.attn_head_size),
+            bias_axes="de",
+            **self._get_common_kwargs_for_sublayer(use_bias=True),
+            name="key",
+        )
+        self._value_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_heads, self.attn_head_size),
+            bias_axes="de",
+            **self._get_common_kwargs_for_sublayer(use_bias=True),
+            name="value",
+        )
+
+        # Relative attention.
+        self._position_dropout_layer = keras.layers.Dropout(self.dropout)
+
+        self._attn_dropout_layer = keras.layers.Dropout(
+            self.dropout, name="attention_dropout"
+        )
+        self._softmax = keras.layers.Softmax(axis=-1, name="attention_softmax")
+
+        # Output.
+        self._output_dense = keras.layers.EinsumDense(
+            equation="abc,cd->abd",
+            output_shape=(None, self.hidden_dim),
+            bias_axes="d",
+            **self._get_common_kwargs_for_sublayer(use_bias=True),
+            name="attention_output",
+        )
+
+    def _get_common_kwargs_for_sublayer(self, use_bias=True):
+        common_kwargs = {}
+
+        kernel_initializer = clone_initializer(self._kernel_initializer)
+        bias_initializer = clone_initializer(self._bias_initializer)
+
+        common_kwargs["kernel_initializer"] = kernel_initializer
+        if use_bias:
+            common_kwargs["bias_initializer"] = bias_initializer
+
+        return common_kwargs
+
+    def _masked_softmax(self, attention_scores, attention_mask=None):
+        """Normalizes the attention scores to probabilities using softmax."""
+
+        if attention_mask is not None:
+            # The expand dim happens starting from the `num_heads` dimension,
```

Comment:
this comment reads somewhat awkward, should we say something more like "Add dimensions so the attention_mask rank equals the attention_scores rank, starting with the num_head dimension" or something like that?

Reply:
I've actually just lifted this from Keras MHA layer: https://github.com/keras-team/keras/blob/master/keras/layers/attention/multi_head_attention.py#L478-L492. Maybe, I'll remove all comments and add one comment, saying that this has been taken from MHA layer.